docs/datastorage/capabilities.md

```diff
@@ -0,0 +1,176 @@
+In order to expose a capability, you will first need an instance of the underlying capability type. There's two ways to obtain such an instance, through the Capability itself, or by explicitly instantiating an implementation of it.
+
+The first method is designed to use a default implementation, if those default values are useful for you. In the case of the Item Handler capability, the default implementation will expose a single slot inventory, which is most probably not what you want.
+
+The second method can be used to customize the existing implementations, or to implement your own. In the case of `IItemHandler`, the default implementation uses the `ItemStackHandler` class, which has an optional argument in the constructor, to specify a number of slots.
+
+```Java
+// A dispenser-like block could use something like this
+IItemHandler inventory = new ItemStackHandler(9);
+```
+
+Once you have your own instance of the capability interface, you will want to notify users of the capability system that you expose this capability. This is done by overriding the `hasCapability` method, and comparing the instance with the capability you are exposing. If your machine has different slots based on which side is being queried, you can test this with the `facing` parameter. For entities and ItemStacks, this parameter can be ignored. Don't forget to fall back to `super`, otherwise the attached capabilities will stop working.
+
+```Java
+@Override
+public boolean hasCapability(Capability<?> capability, EnumFacing facing)
+{
+  if (capability == CapabilityItemHandler.ITEM_HANDLER_CAPABILITY)
+    return true;
+  return super.hasCapability(capability, facing);
+}
+```
+
+Similarly, you will want to provide the interface reference to your capability instance, when requested. Again, don't forget to fall back to `super`.
+
+```Java
+@Override
+public <T> T getCapability(Capability<T> capability, EnumFacing facing)
+{
+  if (capability == CapabilityItemHandler.ITEM_HANDLER_CAPABILITY)
+    return (T)inventory;
+  return super.getCapability(capability, facing);
+}
+```
+
+Attaching Capabilities
+----------------------
+
+As mentioned, attaching capabilities to entities and itemstacks can be done through the AttachCapabilityEvent. This event has 3 sub-events, for extra granularity:
+* `AttachCapabilityEvent.Entity`: Fires only for entities.
+* `AttachCapabilityEvent.TileEntity`: Fires only for tile entities.
+* `AttachCapabilityEvent.Item`: Fires only for item stacks.
+
+In all cases, the event has a method `addCapability`, which can be used to attach capabilities to the target object. Instead of adding capabilities themselves to the list, you add capability providers, which have the chance to return capabilities only from certain sides. While the provider only needs to implement `ICapabilityProvider`, it is good idea to implement `ICapabilitySerializable<T extends NBTBase>`, which on top of returning the capabilities, will allow providing NBT save/load functions.
+
```

Comment:
"If you need to store object specific data it is a good idea to implement"
Because unless you need to store data, implementing that interface does nothing but waste cpu cycles. So don't do it unless you need it.